consensus/src/liveness/proposal_generator.rs

```diff
@@ -135,73 +128,7 @@ impl ProposalGenerator {
-                        WaitingSuccess::WaitWasRequired {
-                            current_duration_since_epoch,
-                            wait_duration,
-                        } => {
-                            counters::PROPOSAL_SUCCESS_WAIT_S.observe_duration(wait_duration);
-                            counters::PROPOSALS_GENERATED_COUNT
-                                .with_label_values(&["wait_was_required"])
-                                .inc();
-                            current_duration_since_epoch
-                        }
-                        WaitingSuccess::NoWaitRequired {
-                            current_duration_since_epoch,
-                            ..
-                        } => {
-                            counters::PROPOSAL_SUCCESS_WAIT_S.observe_duration(Duration::new(0, 0));
-                            counters::PROPOSALS_GENERATED_COUNT
-                                .with_label_values(&["no_wait_required"])
-                                .inc();
-                            current_duration_since_epoch
-                        }
-                    }
-                }
-                Err(waiting_error) => {
-                    match waiting_error {
-                        WaitingError::MaxWaitExceeded => {
-                            counters::PROPOSAL_FAILURE_WAIT_S.observe_duration(Duration::new(0, 0));
-                            counters::PROPOSALS_GENERATED_COUNT
-                                .with_label_values(&["max_wait_exceeded"])
-                                .inc();
-                            bail!(
-                                "Waiting until parent block timestamp usecs {:?} would exceed the round duration {:?}, hence will not create a proposal for this round",
-                                hqc.certified_block().timestamp_usecs(),
-                                round_deadline);
-                        }
-                        WaitingError::WaitFailed {
-                            current_duration_since_epoch,
-                            wait_duration,
-                        } => {
-                            counters::PROPOSAL_FAILURE_WAIT_S.observe_duration(wait_duration);
-                            counters::PROPOSALS_GENERATED_COUNT
-                                .with_label_values(&["wait_failed"])
-                                .inc();
-                            bail!(
-                                "Even after waiting for {:?}, parent block timestamp usecs {:?} >= current timestamp usecs {:?}, will not create a proposal for this round",
-                                wait_duration,
-                                hqc.certified_block().timestamp_usecs(),
-                                current_duration_since_epoch);
-                        }
-                    };
-                }
-            }
-        };
+        let block_timestamp = self.time_service.get_current_timestamp();
```

Comment:
what was the reason for waiting for proposing a block before?

Reply:
(tests?)

Reply:
because the timestamp has to be monotonically increasing, and we don't wait when insert block before. so there's a chance local timestamp < proposed block's parent timestamp.

now we ensure local time past every inserted block's timestamp

Reply:
If a node accepts a block and then the clock goes backward they might accidentally propose a block with a timestamp < the previous block.  We should ensure that doesn't happen, please add a check and fail to propose due to an issue with the clock running backward.

Given it's not intuitive, let's add a comment here, something like:

"All proposed blocks in a branch are guaranteed to have increasing timestamps since their predecessor block will not be added to the BlockStore until the local time exceeds it."